### ISSUE TITLE
bash seems to cache the un-set TZ inside nix-shell

### ISSUE THREAD

**Issue by abathur** [NixOS/nix#2991]
This may be a no-op, but I stumbled into a bash quirk while using nix-shell and wanted to report here (I've reported to bash already) since there's an outside chance it could be causing subtle issues for a small subset of uses of nix-shell.

Basically, an explicitly-set timezone will remain operative after you:

- explicitly set the TZ environment variable to some timezone
- run any external command or builtin that invokes date APIs (date, ls, stat, printf, etc...)
- `unset TZ` (this is what the script for nix-shell will do if TZ was not set at the time of invocation)

I noticed this when I ran the printf builtin with a "%(...)T" strftime format. I'm not sure if it's the only builtin directly affected. As best I can tell, the timezone will stay cached until you do one of two things:

- run any external command/script (I assume bash either does something explicit, or that this is an intrinsic feature of the lower-level APIs it uses to run an external command?). I only used date in the examples below to show that it changes during the command run and not after, but I tried a dozen commands and directly running an empty external script--all triggered the timezone update.
- re-set the TZ env and run any builtin (can be as simple as `TZ=:/etc/localtime :`). The final example below uses this method.

```
$ nix-shell -p coreutils --pure --command "printf '%(%+)T\n';date;printf '%(%+)T\n'"
Sat Jul 13 20:06:19 UTC 2019
Sat Jul 13 15:06:19 CDT 2019
Sat Jul 13 15:06:19 CDT 2019

$ nix-shell -p coreutils --command "printf '%(%+)T\n';date;printf '%(%+)T\n'"
Sat Jul 13 20:06:36 UTC 2019
Sat Jul 13 15:06:36 CDT 2019
Sat Jul 13 15:06:36 CDT 2019

$ nix-shell -p coreutils --command "printf '%(%+)T\n';TZ=:/etc/localtime :;printf '%(%+)T\n'"
Sat Jul 13 20:16:23 UTC 2019
Sat Jul 13 15:16:23 CDT 2019
```

At first I thought this might effect nix-shell shebang scripts, but after poking at it I _think_ they are roughly protected by being wrapped in an exec call. I would guess most other nix-shell invocations have an external command/script execution pretty early, but anything that tries to stick to builtins (including `source`!) for performance reasons could have some subtle issues.

**Comment by abathur** [comment]
I spotted what is presumably the fix for this in the `bash-5.0-release` -> `bash-5.1-alpha` [changenotes](https://tiswww.case.edu/php/chet/bash/CHANGES):
> nn. Fixed a bug that didn't reset the timezone information correctly when the TZ variable was unset.

There's a PR for 5.1 in https://github.com/NixOS/nixpkgs/pull/107354, but I guess this isn't really a non-issue until that's the default... someday :)

**Comment by abathur** [comment]
Bash 5.1 became default in nixpkgs in https://github.com/NixOS/nixpkgs/commit/bdda9a2ed5463, which has now been released for a few major cycles (since 21.11). I guess it's fair to close this even if there'll technically still be installs exposed to it for a while...
